pkg/edgeview/src/basics.go

```diff
@@ -0,0 +1,463 @@
+	opts := strings.Split(opt, ",")
+	for _, o := range opts {
+		ok := false
+		if strings.Contains(o, "/") {
+			opt1 := strings.Split(o, "/")
+			ok = isValidOpt(opt1[0], optslice)
+		} else {
+			ok = isValidOpt(o, optslice)
+		}
+		if !ok {
+			return []string{}, fmt.Errorf("options available: %v", optslice)
+		}
+	}
+	return opts, nil
+}
+
+func isValidOpt(value string, optslice []string) bool {
+	for _, opt := range optslice {
+		if opt == value {
+			return true
+		}
+	}
+	return false
+}
+
+func getBasics() {
+	if runOnServer {
+		if _, err := os.Stat("/config"); err != nil {
+			return
+		}
+	}
+
+	retStr, err := runCmd("ip -4 -br a | grep UP", false, false)
+	if err == nil {
+		lines := splitLines(retStr)
+		ips := []string{}
+		for _, l := range lines {
+			if strings.HasPrefix(l, "bn") {
+				continue
+			}
+			//fmt.Printf("l: %s\n", l)
+			words := strings.Fields(l)
+			n := len(words)
+			ipaddrs := strings.Split(words[n-1], "/")
+			ips = append(ips, ipaddrs[0])
+		}
+		fmt.Printf("Device IPs: %v\n", ips)
+	}
+	retStr, err = runCmd("cat /config/uuid", false, false)
+	if err == nil {
+		fmt.Printf("  UUID: %s", retStr)
+	}
+	retStr, err = runCmd("cat /config/server", false, false)
+	if err == nil {
+		var printed bool
+		conts := strings.Split(retStr, "zedcloud.")
+		if len(conts) == 2 {
+			cont2s := strings.Split(conts[1], ".zededa.net")
+			if len(cont2s) == 2 {
+				cluster := cont2s[0]
+				controller := strings.Replace(retStr, cluster, "\033[0;93m"+cluster+"\033[0m", 1)
+				fmt.Printf("  Controller: %s", controller)
+				printed = true
+			}
+		}
+		if !printed {
+			fmt.Printf("  Controller: %s", retStr)
+		}
+	}
+	retStr, err = runCmd("cat /run/eve-release && echo -n ', ' && cat /run/eve.id", false, false)
+	if err == nil {
+		fmt.Printf("  EVE-OS release %s", retStr)
+	}
+	fmt.Printf("  Edge-View Ver: %s\n", edgeViewVersion)
+	retStr, err = runCmd("uptime", false, false)
+	if err == nil {
+		ret1 := strings.Split(retStr, " load")
+		uptime := ret1[0]
+		loc, _ := time.LoadLocation("UTC")
+		fmt.Printf("  %v(%v), uptime %s", time.Now().In(loc).Format(time.RFC3339), loc, uptime)
+	}
+	fmt.Println()
+	fmt.Println()
+	closePipe(true)
+}
+
+// dynamically install package for uncommon options
+func addPackage(programName, pkgName string) error {
+	retStr, err := runCmd("which "+programName, false, false)
+	if err == nil {
+		if len(retStr) > 0 {
+			fmt.Printf("%s\n", retStr)
+		}
+	} else {
+		_, err = runCmd("apk add "+pkgName, false, false)
+		if err != nil {
+			fmt.Printf("%v\n", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func runCmd(cmd string, isEve, isPrint bool) (string, error) {
+	if isEve && !runOnServer {
+		cmd = "eve exec pillar " + cmd + " 2&>1"
+	}
+	var retStr string
+	var retBytes []byte
+	var err error
+	if runOnServer {
+		retBytes, err = exec.Command("sh", "-c", cmd).Output()
+		if err == nil {
+			retStr = string(retBytes)
+		}
+	} else {
+		retStr, err = remoteRun("root", evDevice, sshprivkey, cmd)
```

Comment:
When is this ssh invocation used? Is this code used on the client in addition to being used in EVE? Confused.

Reply:
edge-view has two modes, one is directly using ssh into the device, without the dispatcher thing; the other is the using dispatcher. they share most of the code, only when executing the commands, one is the launch the 'ssh' as in here, the other is above the, which is 'edge-view' "run-on-server" which is local.

Reply:
and this 'ssh' is not the 'ssh' we were discussing ;-)

Reply:
I wonder if we can structure or comment the code to make the different modes more clear.

Reply:
made extra comments in this section.

Reply:
added comments.

Reply:
Wait, what does the ssh to the server assume here? That ssh as root is enabled into the edge-node? That might be useful for old devices which are not running edge-view, but I don't see why we need it going forward.
So do we need to include this in the edge-view code we integrate?